CMakeLists.txt

```diff
@@ -1,23 +1,37 @@
 project(codal-core C CXX)
 
-RECURSIVE_FIND_DIR(INCLUDE_DIRS "./inc" "*.h")
-RECURSIVE_FIND_FILE(SOURCE_FILES "./source" "*.c??")
+# Set all our default version values
+set( CODAL_VERSION_MAJOR  -1 )
+set( CODAL_VERSION_MINOR  -1 )
+set( CODAL_VERSION_PATCH  -1 )
+set( CODAL_VERSION_BRANCH "master" )
 
-execute_process(WORKING_DIRECTORY "." COMMAND "git" "log" "--pretty=format:%h" "-n" "1" OUTPUT_VARIABLE git_hash)
-execute_process(WORKING_DIRECTORY "." COMMAND "git" "rev-parse" "--abbrev-ref" "HEAD" OUTPUT_VARIABLE git_branch OUTPUT_STRIP_TRAILING_WHITESPACE)
+# Grab the latest tag, hash and branch from git
+execute_process( WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/../${codal.target.name}" COMMAND git describe --tags --abbrev=0 OUTPUT_VARIABLE git_tag_string )
+execute_process( WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/../${codal.target.name}" COMMAND "git" "rev-parse" "--abbrev-ref" "HEAD" OUTPUT_VARIABLE git_branch OUTPUT_STRIP_TRAILING_WHITESPACE)
+execute_process( WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/../${codal.target.name}" COMMAND "git" "log" "--pretty=format:%h" "-n" "1" OUTPUT_VARIABLE git_hash )
 
-if ("${git_branch}" STREQUAL "master")
-    set(CODAL_VERSION_STRING "${CODAL_VERSION_STRING}")
-else()
-    set(CODAL_VERSION_STRING "${CODAL_VERSION_STRING}-${git_branch}-g${git_hash}")
+string( REGEX MATCHALL "[0-9]+" git_tag "${git_tag_string}" )
+
+# If we had a valid git response, update the values
+list(LENGTH git_tag len)
+if( ${len} GREATER 0 )
+    list(GET git_tag 0 CODAL_VERSION_MAJOR )
+    list(GET git_tag 1 CODAL_VERSION_MINOR )
+    list(GET git_tag 2 CODAL_VERSION_PATCH )
 endif()
 
-set(CODAL_VERSION_FLAGS "-DCODAL_VERSION=\\\"${CODAL_VERSION_STRING}\\\"")
+if (NOT "${git_branch}" STREQUAL "master")
+    set(CODAL_VERSION_BRANCH "${git_branch}-${git_hash}")
+endif()
 
-set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CODAL_VERSION_FLAGS}")
+configure_file( "${CMAKE_CURRENT_SOURCE_DIR}/inc/core/codal_version.h.in" "${CMAKE_CURRENT_BINARY_DIR}/gen/codal_version.h" )
```

Comment:
To follow naming conventions let's rename it to `CodalVersion.h`

Reply:
After looking at the rest of the files in this folder, the original name is a better option:
<img width="287" alt="image" src="https://github.com/lancaster-university/codal-core/assets/29712657/aedd2fb1-6b89-45e4-ba1d-95a6a70e5285">